### PR TITLE
Feat/support encoder model (like bert)

### PR DESCRIPTION
**Pull request by woodx9**
## Motivation

Feat/support encoder model (like bert)
## Modifications
1. add bertModel in support models 
2. support bert attention_backend like native torch, triton 
3. add non causal mask in trition extend 

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 1]
```diff
@@ -157,6 +158,10 @@ def forward_extend(
 
         page_table = metadata.page_table
 
+        causal = True
+        if layer.is_cross_attention or layer.attn_type == AttentionType.ENCODER_ONLY:
```
Comment (ispobock) [inline python/sglang/srt/layers/attention/flashattention_backend.py]: Cross attention seems only for flashinfer backend. Don't need to add this condition for flashattention and triton backend.

File: python/sglang/srt/layers/attention/triton_ops/extend_attention.py [hunk 2]
```diff
@@ -94,7 +95,11 @@ def _fwd_kernel(
     offs_d = tl.arange(0, BLOCK_DMODEL)
     offs_dv = tl.arange(0, BLOCK_DV)
     offs_m = tl.arange(0, BLOCK_M)
-    mask_m = (cur_block_m * BLOCK_M + offs_m) < cur_seq_len_extend
+
+    if IS_CAUSAL:
```
Comment (ispobock) [inline python/sglang/srt/layers/attention/triton_ops/extend_attention.py]: Don't need to modify here? `cur_seq_len_extend = cur_seq_len` for encoder model.

File: python/sglang/srt/layers/attention/triton_ops/extend_attention.py [hunk 3]
```diff
@@ -128,7 +133,12 @@ def _fwd_kernel(
 
     for start_n in range(0, cur_seq_len_prefix, BLOCK_N):
         start_n = tl.multiple_of(start_n, BLOCK_N)
-        mask_n = (start_n + offs_n) < cur_seq_len_prefix
+
```
Comment (ispobock) [inline python/sglang/srt/layers/attention/triton_ops/extend_attention.py]: Don't need to modify here? There is no prefix part.

File: python/sglang/srt/layers/attention/triton_ops/extend_attention.py [hunk 4]
```diff
@@ -199,7 +209,11 @@ def _fwd_kernel(
     cur_block_m_end = tl.minimum(cur_seq_len_extend, (cur_block_m + 1) * BLOCK_M)
```
Comment (ispobock) [inline python/sglang/srt/layers/attention/triton_ops/extend_attention.py]: `cur_block_m_end` should adjusted for non causal case. ref: https://github.com/sgl-project/sglang/blob/54b9a2de0a457709607d6df917d5e6ac5004f72b/python/sglang/srt/layers/attention/triton_ops/prefill_attention.py#L95

File: python/sglang/srt/layers/attention/triton_ops/extend_attention.py [hunk 5]
```diff
@@ -243,12 +257,16 @@ def _fwd_kernel(
...
-        else:
+        elif IS_CAUSAL:
             mask_causual = (cur_block_m * BLOCK_M + offs_m[:, None]) >= (
                 start_n + offs_n[None, :]
             )
             mask_causual &= mask_m[:, None] & mask_n[None, :]
             qk = tl.where(mask_causual, qk, float("-inf"))
+        else:
+            mask_non_causal = (cur_block_m * BLOCK_M + offs_m[:, None]) < cur_seq_len
+            mask_non_causal &= mask_m[:, None] & mask_n[None, :]
```
Comment (ispobock) [inline python/sglang/srt/layers/attention/triton_ops/extend_attention.py]: mask_non_causal = mask_m[:, None] & mask_n[None, :]

File: python/sglang/srt/layers/radix_attention.py [hunk 6]
```diff
@@ -18,6 +18,18 @@
 from sglang.srt.model_executor.forward_batch_info import ForwardBatch
 
 
+class AttentionType:
```
Comment (ispobock) [inline python/sglang/srt/layers/radix_attention.py]: Use enum type?

**Comment by woodx9** [comment]
```bash
python -m sglang.launch_server --model-path BAAI/bge-small-en   --disable-radix-cache  --host 0.0.0.0 --is-embedding --chunked-prefill-size -1   --attention-backend torch_native --mem-fraction-static 0.5 --dtype float32
```

have to --disable-radix-cache and  --chunked-prefill-size -1. or some embedding can be wrong.

**Comment by Titan-p** [comment]
LGTM. BTW This unit test is really time-consuming.

**Comment by woodx9** [comment]
Due to embedding accuracy issues,will not use fa3 as the attention backend temporarily. @Titan-p cc

**Review by FlamingoPg — APPROVED** [review]
LGTM to me～

**Review by Titan-p — APPROVED** [review]

**Review by ispobock — APPROVED** [review]

**Review by zhyncs — APPROVED** [review]

**Comment by gabinguo** [comment]
Hola @woodx9 ! 

Is this [line](https://github.com/sgl-project/sglang/blob/6fa6f38ed33e78993ced2b89ed65aebaaaa9aebb/python/sglang/srt/models/bert.py#L337) correct (it's hardcoded to use LAST pooling type)? I change to use CLS pooling and it matches better to the  `transformers` implementation while loading the model `BAAI/bge-large-en-v1.5`.

Here is the comparison between `sglang` embedding endpoint and `transformers` embedding implementation...

(I updated the pooling type to CLS in the codebase)

```
import os
import openai
from transformers import AutoTokenizer, AutoModel
import torch

# --- SGLANG server
# Setup API configuration
os.environ["OPENAI_API_KEY"] = "... a private key"
base_url = "the url that deployed the server to"

client = openai.Client(base_url=base_url)

model = "embedding-model"
sentence = "Something I would like to embed."
# OpenAI SDK for fetching embeddings
response = client.embeddings.create(
    model=model, input=sentence
)
embedding1 = response.data[0].embedding

# --- Native Transformers
# Load model from HuggingFace Hub
tokenizer = AutoTokenizer.from_pretrained('BAAI/bge-large-en-v1.5')
model = AutoModel.from_pretrained('BAAI/bge-large-en-v1.5')
model.eval()

# Tokenize sentences
encoded_input = tokenizer([sentence], padding=True, truncation=True, return_tensors='pt')

# Compute token embeddings
with torch.no_grad():
    model_output = model(**encoded_input)
    # Perform pooling. In this case, cls pooling.
    sentence_embeddings = model_output[0][:, 0]

# normalize embeddings
embedding2 = torch.nn.functional.normalize(sentence_embeddings, p=2, dim=1)[0].tolist()

print("EMB1:\n", embedding1[:5])
print("EMB2:\n", embedding2[:5])
```

---

OUTPUT:

```
EMB1:
 [-0.022556813433766365, -0.019996657967567444, -0.00891366507858038, -0.01984616555273533, 0.01587872952222824]
EMB2:
 [-0.022556817159056664, -0.019996672868728638, -0.008913705125451088, -0.019846146926283836, 0.01587868668138981]
 ```
 
 It's pretty close the two outputs, but still there are some minor difference, Idk how to explain that... do you have a clue?

**Comment by woodx9** [comment]
Hi @gabinguo ,

1. **Pooling Type**  
   Different pooling strategies can be applied to the same embedding architecture.  
   We could consider offering multiple pooling options for users to select based on their specific needs.

2. **Precision Considerations**  
   Which attention backend are you using?  
   Minor discrepancies can occur depending on the backend implementation.  
   A cosine similarity difference on the order of `1e-5` between two vectors is generally acceptable.

**Comment by gabinguo** [comment]
Thanks for the reply @woodx9  : )

1. Sorry to be ignorant, isn't the embedding model e.g. BAAI/bge-large-en-v1.5, fine-tuned with one specific type of pooling strategy. Shouldn't we grab the setting from the pooling type provided in the huggingface model, like [here](https://huggingface.co/BAAI/bge-large-en-v1.5/blob/main/1_Pooling/config.json) ... 

2. I choosed `--attention-backend torch_native` as in the deployment script as you suggested [here](https://github.com/sgl-project/sglang/pull/4887#issuecomment-2766591099)... The difference is way under the order of `1e-5`, it's `3e-10` I think it's negligible.

Thanks again for the explaination!

**Comment by woodx9** [comment]
Hi @gabinguo, thank you for your feedback.

1. **Pooling Type**  
   I will look into whether there is a standardized approach to determine the pooling type from a Hugging Face model.

2. **Attention Backend**  
   You may also consider using `--attention-backend triton`, which offers improved performance. The difference in cosine similarity remains on the order of 1e-5, which is generally negligible.

**Comment by DavidBao03** [comment]
Hi @woodx9 , i'm trying to use Flashinfer backend running bert, but it encounters an error:
```bash
Error: Invalid configuration : NUM_MMA_Q=2 NUM_MMA_D_QK=2 NUM_MMA_D_VO=2 NUM_MMA_KV=4 NUM_WARPS_Q=4 NUM_WARPS_KV=1 please create an issue (https://github.com/flashinfer-ai/flashinfer/issues) and report the issue to the developers.
```
I don't kwon if this error belongs to Flashinfer lib or sglang implementation? Thanks for help!

**Comment by woodx9** [comment]
@DavidBao03 hi, currently only support encoder model with torch_native attn backend and triton attn backend. Other attn backend is not supported yet.

**Comment by DavidBao03** [comment]
Hi @woodx9,  

I was wondering if the pooling type used in BERT should be based on `CLS` rather than `LAST`? Specifically, should it be:  

```python
self.pooler = Pooler(pooling_type=PoolingType.CLS, normalize=True)
```  

When I checked the pooling configurations in [BAAI/bge-small-en](https://huggingface.co/BAAI/bge-small-en/blob/main/1_Pooling/config.json) and [BAAI/bge-base-en-v1.5](https://huggingface.co/BAAI/bge-base-en-v1.5/blob/main/1_Pooling/config.json), both models actually use `CLS` pooling.  

Additionally, I noticed that the test fails when using `BAAI/bge-base-en-v1.5` with `LAST` pooling. 

Thanks for your help!
